numpy/core/src/umath/ufunc_object.c

```diff
@@ -2216,25 +2246,52 @@ PyUFunc_GeneralizedFunction(PyUFuncObject *ufunc,
     /* Get all the arguments */
     retval = get_ufunc_arguments(ufunc, args, kwds,
                 op, &order, &casting, &extobj,
-                &type_tup, &subok, NULL, &axes);
+                &type_tup, &subok, NULL, &axes, &axis, &keepdims);
+    /* if given, need our own reference to axes, as we XDECREF it at the end */
+    Py_XINCREF(axes);
     if (retval < 0) {
         goto fail;
     }
-
+    /*
+     * If keepdims was passed in (and thus changed from the initial value
+     * on top), check the gufunc is suitable.
+     */
+    if (keepdims != -1) {
+        for (i = 0; i < nop; i++) {
+            if (ufunc->core_num_dims[i] != (i < nin? 1 : 0)) {
```

Comment:
I would prefer to see the test separate from the error message. Something like `bool _ufunc_supports_keepdims(ufunc)`

Reply:
Yes, that makes sense. I'll do the same for `axis` (even though it is a trivial one).